pagopa/io-app
# chore: [IAI-109] Added `redux-saga` effect types through `typed-redux-saga`

## Short description
This PR is going to add `typed-redux-saga` to the whole codebase.

**The current PR will add an ESLint rule to lint agains the use of the normal `redux-saga/effects`, at the moment the lint rule is just set to a  _warning_.**

An analysis and an explanation about this PR is described [in this document](https://pagopa.atlassian.net/wiki/spaces/IOAPP/pages/466419879/Analisi+utilizzo+di+typed-redux-saga).

## List of changes proposed in this pull request
- Integrated `typed-redux-saga`
- Integrated `typed-redux-saga` Babel macro to re-integrate `redux-saga` at compile time (this will ensure that the transpiled code will be the same as the current one without this PR)
- Integrated an ESLint rule to avoid `redux-saga/effects` import

## How to test
This refactor _shouldn't_ change the current application's behaviour. 

The changes are only on the dev side, adding a type-check to the Redux-Saga effects. Using a `yield* call(() => true)` should automatically type-check the returned type as `Boolean`.

## Help for the reviewer 
Most of the code is actually auto-generated and the overall changes could be summarized in:
1. Used `typed-redux-saga/macro` instead of `redux-saga/effects`
2. Used `yield*` instead of `yield`
3. Moved the `Effect` type from `redux-saga/effecs` to `ReduxSagaEffect`
4. Moved the current return type of a `take([...])` effect inside the `take` generic (i.e. `take<ActionType<...>>([...])`)
